### PR TITLE
Update Query Caching

### PR DESCRIPTION
**Pull request by jonathangreen**
Makes a few related changes: 
- Updates `hasfulltablecache` to have a cache per database session, so we don't have to worry about invalidating and merging cached items. 
- Updates the query we are using for `_site_configuration_has_changed` to prevent two requests from deadlocking with each other.
- Updates `ExternalIntegration`
  - Not to have a cache, since this cache was never used. 
  - Modify the `lookup` method to make a more sane query when not passing in a library.
  - Modify the relationship to settings so that we don't do a join to `ConfigurationSettings` for every `ExternalIntegrations` query, since we never actually use settings this way. When we have a lot of settings this makes loading an ExternalIntegration very slow, as well as adding a fourth(!) join to the query.

### CODE REVIEW

File: model/configuration.py [hunk 1]
```diff
@@ -382,14 +368,12 @@ def for_collection_and_purpose(cls, _db, collection, purpose):
     @classmethod
     def lookup(cls, _db, protocol, goal, library=None):
 
-        integrations = (
-            _db.query(cls)
-            .outerjoin(cls.libraries)
```
Comment (jonathangreen) [inline model/configuration.py:387]: No need for this join in the common case where we are not filtering on libraries at all.

File: model/__init__.py [hunk 2]
```diff
@@ -417,9 +417,6 @@ def initialize_data(cls, session, set_site_configuration=True):
 
         list(DataSource.well_known_sources(session))
 
-        # Load all existing Genre objects.
-        Genre.populate_cache(session)
```
Comment (jonathangreen) [inline model/__init__.py:421]: Removed since caches are populated on demand now.

File: model/listeners.py [hunk 3]
```diff
@@ -69,7 +69,11 @@ def _site_configuration_has_changed(_db, cooldown=1):
         # Update the timestamp.
         now = utc_now()
         earlier = now - datetime.timedelta(seconds=cooldown)
-        sql = "UPDATE timestamps SET finish=(:finish at time zone 'utc') WHERE service=:service AND collection_id IS NULL AND finish<=(:earlier at time zone 'utc');"
+        sql = (
```
Comment (jonathangreen) [inline model/listeners.py:72]: Need to add a comment here with a bit more detail why we are using a more complicated query to prevent deadlocks.

File: tests/models/test_classification.py [hunk 4]
```diff
@@ -79,104 +79,6 @@ def test_assign_to_genre_can_remove_genre(self):
 
 
 class TestGenre(DatabaseTest):
-    def test_full_table_cache(self):
```
Comment (jonathangreen) [inline tests/models/test_classification.py:82]: These were really cache tests, tacked onto the tests for genre. So I moved them to the fulltablecache test file.

File: model/configuration.py [hunk 5]
```diff
@@ -294,7 +291,6 @@ class ExternalIntegration(Base, HasFullTableCache):
     settings = relationship(
         "ConfigurationSetting",
         backref="external_integration",
-        lazy="joined",
```
Comment (jonathangreen) [inline model/configuration.py:297]: This one liner was causing a lot of chaos in any queries involving `ExternalIntegration` because it caused it to always do a join to ConfigurationSettings and return all the settings for an ExternalIntegration. Adding a join to the query and more data to process. But we never use these settings. They are accessed through `setting()` which doesn't use anything returned in `settings`.

File: model/hasfulltablecache.py [hunk 6]
```diff
@@ -1,140 +1,91 @@
...
+            _db._palace_cache = cache = {}
+        if cls.__name__ not in cache:
+            cache[cls.__name__] = cls.CacheTuple(
+                {}, {}, SimpleNamespace(hits=0, misses=0)
+            )
+        return cache[cls.__name__]
 
     @classmethod
-    def by_id(cls, _db, id):
+    def by_id(cls, _db: Session, id: int) -> Optional[Base]:
```
Comment (vbessonov) [inline model/hasfulltablecache.py:75]: `_cache_lookup` returns a tuple. I think we need to adjust the type hint here as well.

File: model/hasfulltablecache.py [hunk 7]
```diff
@@ -1,140 +1,91 @@
...
 
     def cache_key(self):
         raise NotImplementedError()
 
     @classmethod
-    def _cache_insert(cls, obj, cache, id_cache):
-        """Cache an object for later retrieval, possibly by a different
-        database session.
-        """
+    def _cache_insert(cls, obj: Base, cache: CacheTuple):
```
Comment (vbessonov) [inline model/hasfulltablecache.py:27]: I think it should be `obj: HasFullCachTable` since we're using `cache_key` below in the code.

File: model/hasfulltablecache.py [hunk 8]
```diff
@@ -1,140 +1,91 @@
...
 
     @classmethod
-    def _cache_lookup(cls, _db, cache, cache_name, cache_key, lookup_hook):
+    def _cache_lookup(
+        cls,
+        _db: Session,
+        cache: CacheTuple,
+        cache_name: str,
+        cache_key: Hashable,
+        lookup_hook: Callable,
```
Comment (vbessonov) [inline model/hasfulltablecache.py:41]: Maybe we can rename `lookup_hook` and give it a friendlier name like `cache_miss_hook`?

File: model/hasfulltablecache.py (hunk 6 above)
Comment (jonathangreen) [reply model/hasfulltablecache.py:75]: We call `_cache_lookup` in this function then drop, the new value from the tuple, returning only the object.

File: model/hasfulltablecache.py [hunk 9]
```diff
@@ -1,140 +1,91 @@
...
+            obj, new = lookup_hook()
+            if obj is not None:
+                cls._cache_insert(obj, cache)
+            return obj, new
+
+    @classmethod
+    def get_cache(cls, _db: Session):
+        """Get cache from database session."""
+        try:
+            cache = _db._palace_cache
```
[the remark this answers is not in the document]
Comment (jonathangreen) [reply model/hasfulltablecache.py]: Updated in most recent commit

File: model/hasfulltablecache.py (hunk 7 above)
Comment (jonathangreen) [reply model/hasfulltablecache.py]: Updated to use a protocol, I think this addresses that concern.

File: model/hasfulltablecache.py (hunk 8 above)
Comment (jonathangreen) [reply model/hasfulltablecache.py]: Done

File: model/listeners.py (hunk 3 above)
Comment (jonathangreen) [reply model/listeners.py:72]: Comment added

**Review by tdilauro — CHANGES_REQUESTED** [review]
This looks really good! Requesting a couple of changes:
- Re-enable caching on `ExternalIntegrations`, at least for now.
- Test to ensure that changes made locally are visible through the cache (@vbessonov  is working on this).

File: model/configuration.py [hunk 10]
```diff
@@ -104,7 +104,7 @@ class ExternalIntegrationLink(Base, HasFullTableCache):
     COLLECTION_MIRROR_SETTINGS = settings
 
 
-class ExternalIntegration(Base, HasFullTableCache):
+class ExternalIntegration(Base):
```
[the remark this answers is not in the document]
Comment (jonathangreen) [reply model/configuration.py:107]: Sure. No problem to add it back in. No code in the code base currently called `by_id` to get any items from this cache, so it didn't seem necessary to have it there. Do the improvements you are working on actually use `by_id` to pull in `ExternalIntegrations`?

[the remark this answers is not in the document]
Comment (tdilauro) [reply model/configuration.py:107]: Actually, I think you can safely leave it out. I was keeping my options open, but I'm looking in another direction now. I can add it back, if I need it.

File: model/hasfulltablecache.py [hunk 11]
```diff
@@ -1,140 +1,119 @@
...
-            _db, cls._id_cache, "_id_cache", id, lookup_hook
-        )
+        obj, _ = cls._cache_lookup(db, cache, "id", id, lookup_hook)
         return obj
 
     @classmethod
-    def by_cache_key(cls, _db, cache_key, lookup_hook):
-        return cls._cache_lookup(_db, cls._cache, "_cache", cache_key, lookup_hook)
+    def by_cache_key(
+        cls, db: Session, cache_key: Hashable, lookup_hook: Callable
```
[the remark this answers is not in the document]
Comment (jonathangreen) [reply model/hasfulltablecache.py]: Switched to use `cache_miss_hook`. I like the consistency.

File: model/listeners.py [hunk 12]
```diff
@@ -69,7 +69,22 @@ def _site_configuration_has_changed(_db, cooldown=1):
...
         now = utc_now()
         earlier = now - datetime.timedelta(seconds=cooldown)
-        sql = "UPDATE timestamps SET finish=(:finish at time zone 'utc') WHERE service=:service AND collection_id IS NULL AND finish<=(:earlier at time zone 'utc');"
+
+        # We use SKIP LOCKED in the query here, to skip this update if any other
+        # process is currently updating the timestamp. This prevents us from waiting
+        # for the other transaction to update the timestamp, only to update it ourselves
+        # it also avoids a possible deadlock where a request calls back into itself
+        # for example during the library registration process.
+        #
```
Comment (tdilauro) [inline model/listeners.py:78]: Minor: looks like two sentences have run together here. Suggest rewording:

Using SKIP LOCKED here allows us avoid waiting for another process that is (presumably) already updating the timestamp, only to immediately update it again ourselves after the wait. It also avoids a possible deadlock for cases in which a request results in another call into the app server while that request is still active. For example, during library registration...

**Review by tdilauro — COMMENTED** [review]
Added a few more minor suggestions, but nothing that would stop merging. Just need to get the tests are sorted out.

File: lane.py [hunk 13]
```diff
@@ -3202,9 +3202,13 @@ def configuration_relevant_update(mapper, connection, target):
...
-        Library._has_root_lane_cache.clear()
+@event.listens_for(Lane.library_id, "set")
+@event.listens_for(Lane.root_for_patron_type, "set")
+def receive_modified(target, value, oldvalue, initiator):
+    # Some elements of Lane configuration are stored in the
+    # corresponding Library objects for performance reasons.
+
+    # Remove this information whenever the Lane configuration
+    # changes. This will force it to be recalculated.
+    Library._has_root_lane_cache.clear()
```
Comment (jonathangreen) [inline lane.py:3214]: This fixes the failing `Lane` test. It appears that before because we were clearing the table cache during a session, we were causing items to be flushed more often and this made the "after_update" hook fire. 

It only fires after a flush to the DB. The problem with this cache was that items can be stale, without a flush happening. Hooking this set event clears this up, so that the cache is cleared when we need it to be. I'm not sure how the changes in this PR necessitated this change though.

File: model/listeners.py (hunk 12 above)
Comment (jonathangreen) [reply model/listeners.py]: 👌

**Review by tdilauro — APPROVED** [review]
Looks good! 🎸 

Approving this now, but it might be good to check into [this failing check](https://github.com/ThePalaceProject/circulation-core/runs/4493276200?check_suite_focus=true) from the commit [9d10704](https://github.com/ThePalaceProject/circulation-core/pull/44/commits/9d107043e95da00cac5425f7a6443b35275539d0)

**Comment by tdilauro** [comment]
Noting here that I added a commit with an elapsed time logging context manager.

File: util/log.py [hunk 14]
```diff
@@ -22,3 +32,26 @@ def wrapper(*args, **kwargs):
...
+    """Context manager for logging elapsed time.
+
+    :param log_method: Callable to be used to log the message(s).
+    :param message_prefix: Optional string to be prepended to the emitted log records.
+    :param skip_start: Boolean indicating whether to skip the starting message.
+    """
+
+    prefix = f"{message_prefix}: " if message_prefix else ""
+    if not skip_start:
+        log_method(f"{prefix}Starting...")
```
Comment (vbessonov) [inline util/log.py:50]: If we have a prefix, then it probably makes sense to use a lower case of `Starting...`

**Review by vbessonov — APPROVED** [review]
